### DISCUSSION TITLE
Finding the current refinement level of a MeshBlock

### DISCUSSION THREAD

**Discussion by hwhitehead** [PrincetonUniversity/athena#494] [Q&A] [answered]
Is it possible to access the refinement level of a MeshBlock from its pointer? I am aware of a subclass LogicalLocation which I thought might contain details of the refinement level, but the int values stored in `pmb->loc.level` seem to be too large to represent levels of refinement (perhaps levels within the MeshBlock tree?). I could not find anything in the MeshRefinement class either. 

I can calculate the refinement level by comparing the MeshBlock size to the root level, but this seems overly expensive.

**Comment by c-white** [comment]
The levels you found are indeed levels in the tree, and they are offset by `root_level` from the ones you want. Try `pmb->loc.level - pmb->pmy_mesh->root_level`.

**Reply by hwhitehead** [reply]
Thanks Chris, appreciate it

**Reply by hwhitehead** [reply]
Unfortunately it seems that root_level is privatised within the Mesh class. Is there any alternative way to access this value?

**Reply by c-white** [reply, marked as answer]
Any `Mesh::` functions (like `InitUserMeshData`), as well as `MeshBlock::` functions (like `ProblemGenerator`) should be able to see the value. I guess it's not readily accessible to any user-defined source function.

In that case the easiest thing to do is move the variable declaration in `mesh.hpp` to the public section of the class definition. Or just delete the `private:` statement. The code will work fine with all variables being public.

The fact that you want to access something so deep and uncommon is a little bit of a warning flag. Physics should depend on physical quantities like coordinates, rather than code bookkeeping quantities like refinement level.

**Reply by hwhitehead** [reply]
Understood, I wasn't sure if publicising this value would cause issues elsewhere.

I acknowledge the oddity of this problem. It is somewhat particular to my need to finely resolve around a body that moves throughout the domain in time. Were the body static, SMR or an irregular mesh would easily solve the problem. Thanks for the help.
